applications/visualizer/backend/tests/ingestion/test_validator.py

```diff
@@ -0,0 +1,463 @@
+                name="ADAL",
+                emb=True,
+                nt="l",
+                intail=False,
+                classes="ADA",
+                typ="i",
+            ),
+            Neuron(
+                inhead=True,
+                name="ADAR",
+                emb=True,
+                nt="l",
+                intail=False,
+                classes="ADA",
+                typ="i",
+            ),
+        ],
+        datasets=[
+            Dataset(
+                id="white_1986_jse",
+                name="White et al., 1986, JSE (adult)",
+                type=DatasetType.TAIL,
+                time=60,
+                visualTime=50,
+                description="Adult legacy tail with pre-anal ganglion",
+            )
+        ],
+        connections={
+            "white_1986_jse": [
+                Connection(
+                    ids=[6949667],
+                    post="ADAR",
+                    post_tid=[6693872],
+                    pre="ADAL",
+                    pre_tid=[6719648],
+                    syn=[1],
+                    typ=ConnectionType.CHEMICAL,
+                )
+            ]
+        },
+        annotations={
+            "head": Annotation(
+                root={
+                    "increase": [
+                        ("ADAL", "RIPL"),
+                        ("ADAR", "RIPR"),
+                        ("ADEL", "AVKR"),
+                    ],
+                    "postembryonic": [
+                        ("ADAL", "RMFL"),
+                    ],
+                }
+            ),
+            "complete": Annotation(),
+        },
+    ),
+]
+
+
+@pytest.mark.parametrize("data_fn", instanciate_valid_data_tc)
+def test__valid_data(data_fn: Callable[..., Data]):
+    # every data definition should be able to correctly
+    # be instanciated
+    data_fn()
+
+
+instanciate_invalid_data_tc: List[Callable[..., Data]] = [
+    lambda: Data(
+        neurons=[],
+        datasets=[],
+        connections={"white_1986_jse": []},  # dataset id not in datasets
+    ),
+    lambda: Data(
+        neurons=[],
+        datasets=[],
+        connections={},
+        annotations={
+            # headd is not a valid annotation key
+            "headd": Annotation(),  # type: ignore
+        },
+    ),
+]
+
+
+@pytest.mark.parametrize("data_fn", instanciate_invalid_data_tc)
+def test__invalid_data(data_fn: Callable[..., Data]):
+    # every data definition should raise an exception upon
+    # being instanciated
+    with pytest.raises(ValidationError):
+        data_fn()
+
+
+@pytest.fixture
+def data_fixture(request: pytest.FixtureRequest) -> JSON:
+    # TODO: can we move fixtures close to the test? i.g Path(request.fspath).parent / Path("fixtures")
+    ROOT_DIR = Path(request.fspath).parent.parent.parent.parent.parent.parent  # type: ignore
+    FIXTURES_DIR = ROOT_DIR / Path("data/db-raw-data")
+
+    def load_file(f: str) -> Dict:
+        with open(FIXTURES_DIR / Path(f)) as file:
```

Comment:
Working with `Path`, it's better to use the `.open(...)` method on it:

```python
with (FIXTURES_DIR / f).open() as file:
  ...
```

Another way of doing it could be:

```python
json.loads((FIXTURES_DIR / f).read_text())